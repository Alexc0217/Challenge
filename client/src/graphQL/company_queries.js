import { gql } from '@apollo/client';

export const GET_COMPANY = (id) => gql`
  {
    company(id: ${id}){
      name
      id
    }
  }
`

export const CREATE_COMPANY = gql`
  mutation CreateCompany($name: String!) {
    createCompany(input: { name: $name }) {
      company {
        id
        name
      }
      errors
    }
  }
`;

export const COMPANIES = gql`
  {
   companies{
      id
      name
      employeesNumber
    }  
  }
`

export const COMPANY_EMPLOYEES = (id) => gql`
  {
    companyEmployees(companyId: ${id}){
      companyName
      employees{
        name
        id
        role
        email
        managerId
      }
    }
  }
`

export const UPDATE_EMPLOYEE_MANAGER = gql`
  mutation UpdateEmployeeManager($employeeId: ID!, $managerId: ID!) {
    updateEmployeeManager(input: {employeeId: $employeeId, managerId: $managerId}) {
      employee{
        name
        id
      }
      message
      errors
    }
  }
`;