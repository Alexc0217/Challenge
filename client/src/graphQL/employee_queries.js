import { gql } from '@apollo/client';

export const CREATE_EMPLOYEE = gql`
  mutation CreateEmployee($name: String!, $email: String!, $role: String!, $companyId: ID!){
    createEmployee(input: {name: $name, email: $email, role: $role, companyId: $companyId}){
      employee{
        name
        email
        role
      }
      message
      errors
    }
  }
`

export const DELETE_EMPLOYEE = gql`
  mutation DeleteEmployee($id: ID!){
    deleteEmployee(input: {id: $id}){
      employee{
        name
        id
      }
      message
      errors
    }
  }
`

export const GET_EMPLOYEE_PAIRS = (id) => gql`
  {
    employee(id: ${id}){
      name
      id
      role
      pairs{
        name
        id
        role
      }
    }
  }
`

export const GET_EMPLOYEE_SUBORDINATES = (id) => gql`
  {
    employee(id: ${id}){
      name
      id
      role
      subordinates{
        name
        id
        role
      }
    }
  }
`

export const GET_EMPLOYEE_SECOND_LEVEL = (id) => gql`
  {
    employee(id: ${id}){
      name
      subordinates{
        name
        id
        email
        role
        subordinates{
          id
          name
          email
          role
        }
      }
    }
  }
`