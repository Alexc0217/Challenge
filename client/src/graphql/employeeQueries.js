import { gql } from '@apollo/client';

export const CREATE_EMPLOYEE = gql`
  mutation CreateEmployee($name: String!, $email: String!, $role: String!, $companyId: ID!){
    createEmployee(input: {name: $name, email: $email, role: $role, companyId: $companyId}){
      employee{
        name
        email
        role
      }
      message
      errors
    }
  }
`

export const DELETE_EMPLOYEE = gql`
  mutation DeleteEmployee($id: ID!){
    deleteEmployee(input: {id: $id}){
      employee{
        name
        id
      }
      message
      errors
    }
  }
`

export const GET_EMPLOYEE_PAIRS = gql`
  query Employee($id: ID!){
    employee(id: $id){
      name
      id
      role
      pairs{
        name
        id
        role
      }
    }
  }
`

export const GET_EMPLOYEE_SUBORDINATES = gql`
  query Employee($id: ID!){
    employee(id: $id){
      name
      id
      role
      subordinates{
        name
        id
        role
      }
    }
  }
`

export const GET_EMPLOYEE_SECOND_LEVEL = gql`
  query Employee($id: ID!){
    employee(id: $id){
      name
      id
      role
      subordinates{
        name
        id
        email
        role
        subordinates{
          id
          name
          email
          role
        }
      }
    }
  }
`

export const UPDATE_EMPLOYEE_MANAGER = gql`
  mutation UpdateEmployeeManager($employeeId: ID!, $managerId: ID!) {
    updateEmployeeManager(input: {employeeId: $employeeId, managerId: $managerId}) {
      employee{
        name
        id
      }
      message
      errors
    }
  }
`;