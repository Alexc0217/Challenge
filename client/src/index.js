import React from 'react';
import ReactDOM from 'react-dom/client';
import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';
import App from './pages/App';
import { client } from './graphql/client';
import { ApolloProvider } from '@apollo/client';

const root = ReactDOM.createRoot(document.getElementById('root'));
root.render(
  <ApolloProvider client={client}>
    <Router>
      <Routes>
        <Route element={<App />} path="/*" />
      </Routes>
    </Router>
  </ApolloProvider>
);