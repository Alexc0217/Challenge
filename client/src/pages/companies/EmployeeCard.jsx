import { React, useCallback } from "react";
import { Select, MenuItem, FormControl, InputLabel } from "@mui/material";
import { useNavigate } from "react-router-dom";
import { Image } from "../../components/ui/styles";
import UserIcon from "../../assets/images/UserIcon.png"
import { DELETE_EMPLOYEE } from "../../graphQL/employee_queries";
import { useMutation } from '@apollo/client';
import { UPDATE_EMPLOYEE_MANAGER } from "../../graphQL/employee_queries";
import {SwalError, SwalSuccess } from "../../components/ui";
import { EMPLOYEE } from "../../navigation/routes";
import { pathWithParams } from "../../utils/pathWithParams";
import {
  CardEmployee,
  CardRight,
  EmployeePicture,
  EmployeeInfos,
  Label,
  EmployeeActions,
  ActionButton,
} from "./styles";

export default function EmployeeCard({employee, employees}){
  let navigate = useNavigate();

  const [updateManager] = useMutation(UPDATE_EMPLOYEE_MANAGER, {
    onCompleted: (response) => {
      const data = response.updateEmployeeManager;
      
      if(data.errors.length > 0) return SwalError(data.errors);
      SwalSuccess({message: data.message});
    }, 
    onError: (error) => {
      SwalError(error.message);
    }
  });

  const [deleteEmployee] = useMutation(DELETE_EMPLOYEE, {
    onCompleted: (response) => {
      const data = response.deleteEmployee;

      if(data.errors.length > 0) return SwalError(data.errors);
      SwalSuccess({message: data.message});
    },
    onError: (error) => {
      SwalError(error.message);
    }
  })
  
  const handleManagerChange = useCallback((employeeId, newManagerId) => {
    updateManager({ variables: { employeeId, managerId: newManagerId } });
  }, [updateManager]);

  return(
    <CardEmployee key={employee.id}>
      <CardRight>
        <EmployeePicture>
          <Image src={UserIcon}/>
        </EmployeePicture>
        <EmployeeInfos>
          <Label>Nome: {employee.name}</Label>
          <Label>E-mail: {employee.email}</Label>
          <Label>Cargo: {employee.role || "Sem cargo"}</Label>
          <FormControl fullWidth variant="outlined" size="small" style={{ marginTop: '8px' }}>
            <InputLabel>Selecione o Gestor</InputLabel>
            <Select
              label="Selecione o Gestor"
              value={employee.managerId || ''}
              onChange={(e) => handleManagerChange(employee.id, e.target.value)}
            >
              <MenuItem value="">
                <em>None</em>
              </MenuItem>
              {employees
                .filter((emp) => emp.id !== employee.id)
                .map((manager) => (
                  <MenuItem key={manager.id} value={manager.id}>
                    {manager.name}
                  </MenuItem>
                ))}
            </Select>
          </FormControl>
        </EmployeeInfos>
      </CardRight>
      <EmployeeActions>
        <ActionButton variant="contained" size="small" onClick={() => navigate(pathWithParams(EMPLOYEE.PAIRS, {id: employee.id}))}>
          Ver colaboradores do mesmo n√≠vel
        </ActionButton>
        <ActionButton variant="contained" size="small" onClick={() => navigate(pathWithParams(EMPLOYEE.SUBORDINATES, {id: employee.id}))}>
          Ver liderados
        </ActionButton>
        <ActionButton variant="contained" size="small" onClick={() => navigate(pathWithParams(EMPLOYEE.SECOND_LEVEL, {id: employee.id}))}>
          Ver liderados dos liderados
        </ActionButton>
        <ActionButton variant="contained" type="danger" size="small" onClick={() => (deleteEmployee({variables: {id: employee.id}}))}>
          Apagar colaborador
        </ActionButton>
      </EmployeeActions>
    </CardEmployee>
  )
}