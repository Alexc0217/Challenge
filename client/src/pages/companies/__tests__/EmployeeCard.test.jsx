import React from 'react';
import renderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import EmployeeCard from '../EmployeeCard';
import { MemoryRouter } from 'react-router-dom';
import { UPDATE_EMPLOYEE_MANAGER, DELETE_EMPLOYEE } from '../../../graphQL/employeeQueries';

const mocks = [
  {
    request: {
      query: UPDATE_EMPLOYEE_MANAGER,
      variables: { employeeId: '1', managerId: '2' },
    },
    result: {
      data: {
        updateEmployeeManager: { errors: [], message: 'Manager updated successfully' },
      },
    },
  },
  {
    request: {
      query: DELETE_EMPLOYEE,
      variables: { id: '1' },
    },
    result: {
      data: {
        deleteEmployee: { errors: [], message: 'Employee deleted successfully' },
      },
    },
  },
];

const employee = {
  id: '1',
  name: 'Alex Moura',
  email: 'alex.moura@outlook.com',
  role: 'Developer',
  managerId: '2',
};

const employees = [
  { id: '1', name: 'Bruno Souza' },
  { id: '2', name: 'Bruna Souza' },
];

describe('EmployeeCard Component Snapshot', () => {
  it('matches the snapshot', () => {
    const tree = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MemoryRouter>
          <EmployeeCard employee={employee} employees={employees} />
        </MemoryRouter>
      </MockedProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
