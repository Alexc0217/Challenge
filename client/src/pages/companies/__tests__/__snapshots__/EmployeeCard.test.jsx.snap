// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeCard Component Snapshot matches the snapshot 1`] = `
<div
  className="sc-eEPDDI crhsec"
>
  <div
    className="sc-gutikT jLThts"
  >
    <div
      className="sc-huvEkS fotIpO"
    >
      <img
        className="sc-iCZwEW jXrbTg"
        src="UserIcon.png"
      />
    </div>
    <div
      className="sc-fLVwEd iRiZeL"
    >
      <p
        className="sc-gLXSEc chuanv"
      >
        Nome: 
        Alex Moura
      </p>
      <p
        className="sc-gLXSEc chuanv"
      >
        E-mail: 
        alex.moura@outlook.com
      </p>
      <p
        className="sc-gLXSEc chuanv"
      >
        Cargo: 
        Developer
      </p>
      <div
        className="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        style={
          Object {
            "marginTop": "8px",
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={true}
        >
          Selecione o Gestor
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          onClick={[Function]}
        >
          <div
            aria-controls=":r0:"
            aria-expanded="false"
            aria-haspopup="listbox"
            className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            role="combobox"
            tabIndex={0}
          >
            Bruna Souza
          </div>
          <input
            aria-hidden={true}
            aria-invalid={false}
            className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            disabled={false}
            onAnimationStart={[Function]}
            onChange={[Function]}
            required={false}
            tabIndex={-1}
            value="2"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-14lo706"
            >
              <span>
                Selecione o Gestor
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-bSlUec iApYxM"
  >
    <button
      className="sc-jcHdAB fGmA-Dy"
      onClick={[Function]}
      size="small"
      variant="contained"
    >
      Ver colaboradores do mesmo n√≠vel
    </button>
    <button
      className="sc-jcHdAB fGmA-Dy"
      onClick={[Function]}
      size="small"
      variant="contained"
    >
      Ver liderados
    </button>
    <button
      className="sc-jcHdAB fGmA-Dy"
      onClick={[Function]}
      size="small"
      variant="contained"
    >
      Ver liderados dos liderados
    </button>
    <button
      className="sc-jcHdAB gkgReA"
      onClick={[Function]}
      size="small"
      type="danger"
      variant="contained"
    >
      Apagar colaborador
    </button>
  </div>
</div>
`;
